### DISCUSSION TITLE
Unable to build OpenSSL on Windows after I modified a function

### DISCUSSION THREAD

**Discussion by IonutCorbu** [openssl/openssl#24082] [Q&A] [answered]
Hello! I'm trying to use CMS_SignerInfo_sign function, but I don't want Signing Time attribute to be present. So I tried to recompile OpenSSL latest version with a custom function starting from the body of the specified function:
cms_sd.c
```
int CMS_SignerInfo_sign_custom(CMS_SignerInfo *si,int withSigningTime)
{
    EVP_MD_CTX *mctx = si->mctx;
    EVP_PKEY_CTX *pctx = NULL;
    unsigned char *abuf = NULL;
    int alen;
    size_t siglen;
    const CMS_CTX *ctx = si->cms_ctx;
    char md_name[OSSL_MAX_NAME_SIZE];

    if (OBJ_obj2txt(md_name, sizeof(md_name),
                    si->digestAlgorithm->algorithm, 0) <= 0)
        return 0;

    if (withSigningTime && CMS_signed_get_attr_by_NID(si, NID_pkcs9_signingTime, -1) < 0) {
        if (!cms_add1_signingTime(si, NULL))
            goto err;
    }

    if (!ossl_cms_si_check_attributes(si))
        goto err;

    if (si->pctx) {
        pctx = si->pctx;
    } else {
        EVP_MD_CTX_reset(mctx);
        if (EVP_DigestSignInit_ex(mctx, &pctx, md_name,
                                  ossl_cms_ctx_get0_libctx(ctx),
                                  ossl_cms_ctx_get0_propq(ctx), si->pkey,
                                  NULL) <= 0)
            goto err;
        si->pctx = pctx;
    }

    alen = ASN1_item_i2d((ASN1_VALUE *)si->signedAttrs, &abuf,
                         ASN1_ITEM_rptr(CMS_Attributes_Sign));
    if (!abuf)
        goto err;
    if (EVP_DigestSignUpdate(mctx, abuf, alen) <= 0)
        goto err;
    if (EVP_DigestSignFinal(mctx, NULL, &siglen) <= 0)
        goto err;
    OPENSSL_free(abuf);
    abuf = OPENSSL_malloc(siglen);
    if (abuf == NULL)
        goto err;
    if (EVP_DigestSignFinal(mctx, abuf, &siglen) <= 0)
        goto err;

    EVP_MD_CTX_reset(mctx);

    ASN1_STRING_set0(si->signature, abuf, siglen);

    return 1;

 err:
    OPENSSL_free(abuf);
    EVP_MD_CTX_reset(mctx);
    return 0;
}
```

and I also added the signature of the function in cms.h.in:
```
int CMS_SignerInfo_sign_custom(CMS_SignerInfo *si,int withSigningTime);
```
I used the following commands:
```
perl Configure VC-WIN64A --prefix="C:\Program Files\Openssl_built" no-shared
nmake
nmake test
nmake install
```
After it is done, I tried to link the compiled version with my project, but when I call it, I have a series of "unresolved references":
![image](https://github.com/openssl/openssl/assets/85988309/9bd8fcfa-27c7-453f-a645-a94d455f8db7)

How do I have to compile OpenSSL on Windows if I added a function? 
I tried the same thing on Linux respecting commands from [Hacking.md](https://github.com/openssl/openssl/blob/master/HACKING.md) and it is working for Linux.

**Comment by IonutCorbu** [comment]
I tried to add my prototype of function also in libcrypto.num as it follows:
```
CMS_SignerInfo_sign_custom	1555	3_0_0	EXIST::FUNCTION:CMS
```
but now, my app is crashing because it's not finding the entry point...
![image](https://github.com/openssl/openssl/assets/85988309/2dfb0bb8-54f6-4e1b-8230-6c4cf6c5ba8d)

I don't understand why and how do i have to compile OpenSSL after I added a function on Windows.

Do you have any idea why is this happening?

**Comment by beldmit** [comment]
I'm afraid you will have to run `make update` but not sure it will work for Windows

**Reply by IonutCorbu** [reply, marked as answer]
make update does not have an equivalent in Windows like 
```
nmake update
```
But I finally solved the problem by adding an entry in libcrypto.num with my function something like:
```
CMS_SignerInfo_sign_custom  ? 3_4_0        EXIST::FUNCTION:
```
And in Hacking.md file it is specified the same action done by make update (libcrypto.num is updated).
